### PR TITLE
Implementation of the loss from Supervised Contrastive Loss paper. 

### PR DESCRIPTION
**Pull request by Nimrais**
## Before submitting (checklist)

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)?
- [x] Did you check the code style? `catalyst-make-codestyle && catalyst-check-codestyle` (`pip install -U catalyst-codestyle`).
- [x] Did you make sure to update the docs? We use Google format for all the methods and classes.
- [ ] Did you check the docs with `make check-docs`?
- [x] Did you write any new necessary tests?
- [x] Did you check that your code passes the unit tests `pytest .` ? 
- [ ] Did you add your new functionality to the docs?
- [x] Did you update the [CHANGELOG](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md)?
- [ ] Did you run [colab minimal CI/CD](https://colab.research.google.com/github/catalyst-team/catalyst/blob/master/examples/notebooks/colab_ci_cd.ipynb) with `latest` and `minimal` requirements?
- [ ] Did you check XLA integration with [single](https://colab.research.google.com/github/catalyst-team/catalyst/blob/master/examples/notebooks/Catalyst_XLA_single_process.ipynb) and [multiple](https://colab.research.google.com/github/catalyst-team/catalyst/blob/master/examples/notebooks/Catalyst_XLA_multi_process.ipynb) processes?

## Description

In the paper Supervised Contrastive Loss, the abstract contrastive criterion was proposed.
This PR carefully implements this criterion. 

## Type of Change

- [x] Examples / docs / tutorials / contributors update
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Improvement (non-breaking change which improves an existing feature)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### FAQ
Please review the FAQ before submitting an issue:
- [x] I have read the [documentation and FAQ](https://catalyst-team.github.io/catalyst/)
- [x] I have reviewed the [minimal examples section](https://github.com/catalyst-team/catalyst#minimal-examples)
- [x] I have checked the [changelog](https://github.com/catalyst-team/catalyst/blob/master/CHANGELOG.md) for main framework updates
- [x] I have read the [contribution guide](https://github.com/catalyst-team/catalyst/blob/master/CONTRIBUTING.md)
- [x] I have joined [Catalyst slack (#__questions channel)](https://join.slack.com/t/catalyst-team-core/shared_invite/zt-d9miirnn-z86oKDzFMKlMG4fgFdZafw) for issue discussion

### CODE REVIEW

File: examples/catalyst_contrastive/supervised_contrastive.py [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+from torch.optim import Adam
+import torchvision
+
+from catalyst import dl
+from catalyst.contrib import nn
+from catalyst.contrib.models.cv.encoders import ResnetEncoder
+from catalyst.contrib.nn.criterion import NTXentLoss
+from catalyst.contrib.nn.criterion.supervised_contrastive import SupervisedContrastiveLoss
+from catalyst.data.dataset.self_supervised import SelfSupervisedDatasetWrapper
+from catalyst.dl import SelfSupervisedRunner
```
Comment (Scitator) [inline examples/catalyst_contrastive/supervised_contrastive.py]: could we use it as `dl.SelfSupervisedRunner`?

File: examples/catalyst_contrastive/supervised_contrastive.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+import torch.nn.functional as F
+from torch.optim import Adam
+import torchvision
+
+from catalyst import dl
+from catalyst.contrib import nn
+from catalyst.contrib.models.cv.encoders import ResnetEncoder
+from catalyst.contrib.nn.criterion import NTXentLoss
+from catalyst.contrib.nn.criterion.supervised_contrastive import SupervisedContrastiveLoss
+from catalyst.data.dataset.self_supervised import SelfSupervisedDatasetWrapper
```
Comment (Scitator) [inline examples/catalyst_contrastive/supervised_contrastive.py]: could we use it as `data.SelfSupervisedDatasetWrapper`?

**Review by Scitator — DISMISSED** [review]
LGTM 👍 
it also would be great to finally rename `catalyst_contrastive` to `self_supervised` ;)

File: catalyst/contrib/nn/criterion/supervised_contrastive.py [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        Args:
+            features: [bs; feature_len]
+            targets: [bs]
+
+        Returns:
+            computed loss
+        """
+        # torch.exp(1/tau) self similarity
+        self_similarity = 1 / self.tau
+        exp_self_similarity = e ** (1 / self.tau)
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/supervised_contrastive.py]: could we move them under `__init__`? 👀

File: examples/catalyst_contrastive/supervised_contrastive.py (hunk 2 above)
Comment (Nimrais) [reply examples/catalyst_contrastive/supervised_contrastive.py]: [Done](https://github.com/catalyst-team/catalyst/pull/1293/commits/57766500548d68bbf3c7dcaaa9db0525fa3e0179)

File: examples/catalyst_contrastive/supervised_contrastive.py (hunk 1 above)
Comment (Nimrais) [reply examples/catalyst_contrastive/supervised_contrastive.py]: [Done](https://github.com/catalyst-team/catalyst/pull/1293/commits/a476171c68e61ec6864271953a77497c1bd8fc39)

File: catalyst/contrib/nn/criterion/supervised_contrastive.py (hunk 3 above)
Comment (Nimrais) [reply catalyst/contrib/nn/criterion/supervised_contrastive.py]: Do you think it would be better?
I don't have any preferences.

Comment (Scitator) [reply catalyst/contrib/nn/criterion/supervised_contrastive.py]: it should be a bit-bit faster
or at least it should remove batch-independent code from forward pass

Comment (Nimrais) [reply catalyst/contrib/nn/criterion/supervised_contrastive.py]: hmm...
You are right.

Comment (Nimrais) [reply catalyst/contrib/nn/criterion/supervised_contrastive.py]: [Done](https://github.com/catalyst-team/catalyst/pull/1293/commits/792ec6d7d074246423f41678288a846c18865604)

File: catalyst/contrib/nn/criterion/__init__.py [hunk 4]
```diff
@@ -1,5 +1,6 @@
 # flake8: noqa
 
+import catalyst
```
Comment (Scitator) [inline catalyst/contrib/nn/criterion/__init__.py]: why do we need it here?

**Review by Scitator — COMMENTED** [review]
last one comment

File: catalyst/contrib/nn/criterion/__init__.py (hunk 4 above)
Comment (Nimrais) [reply catalyst/contrib/nn/criterion/__init__.py]: [deleted](https://github.com/catalyst-team/catalyst/pull/1293/commits/e3a0fb7348352a40fd041fd0e3135fe1f2f5d3fd)

**Review by Scitator — APPROVED** [review]
